analytics_data_api/v0/models.py

```diff
@@ -394,24 +395,29 @@ class ModuleEngagementTimelineManager(models.Manager):
     def get_timelines(self, course_id, username):
         queryset = ModuleEngagement.objects.all().filter(course_id=course_id, username=username) \
             .values('date', 'entity_type', 'event') \
-            .annotate(count=Sum('count')) \
+            .annotate(total_count=Sum('count')) \
+            .annotate(distinct_entity_count=Count('entity_id')) \
             .order_by('date')
 
         timelines = []
 
-        for key, group in groupby(queryset, lambda x: (x['date'])):
-            # Iterate over groups and create a single item with engagement data
-            item = {
-                u'date': key,
+        for date, engagements in groupby(queryset, lambda x: (x['date'])):
+            # Iterate over engagements for this day and create a single day with
+            # engagement data.
+            day = {
+                u'date': date,
             }
-            for engagement in group:
-                entity_type = engagement_entity_types.SINGULAR_TO_PLURAL.get(engagement['entity_type'],
-                                                                             engagement['entity_type'])
-                engagement_type = '{}_{}'.format(entity_type, engagement['event'])
-                count = item.get(engagement_type, 0)
-                count += engagement['count']
-                item[engagement_type] = count
-            timelines.append(item)
+            for engagement in engagements:
+                engagement_type = EngagementType(engagement['entity_type'], engagement['event'])
+                current_count = day.get(engagement_type.name, 0)
+
+                if engagement_type.is_counted_by_entity:
+                    count_delta = engagement['distinct_entity_count']
```

Comment:
Fair point, though I'd like to leave it as-is for the following reasons, if that's alright with you:
* If that breakage were to happen, our tests would fail.
* As you mention, the line is quite long, and it goes past our edX limit of 120 characters.